File: app/src/main/java/de/troido/bledemo/epd/conversion/SplitColor.java
package de.troido.bledemo.epd.conversion;

import android.graphics.Color;
import android.support.annotation.ColorInt;

/**
 * Color split into channels which can take up negative values for dithering
 * error purposes.
 */
public final class SplitColor {
    private int r;
    private int g;
    private int b;

    public SplitColor(int r, int g, int b) {
        this.r = r;
        this.g = g;
        this.b = b;
    }

    public SplitColor(@ColorInt int color) {
        this(Color.red(color), Color.green(color), Color.blue(color));
    }

    /**
     * Adds the given error color multiplied by the given numerator
     * (with 16 as denominator), mutating this object.
     */
    public SplitColor addErrMut(SplitColor color, int numerator) {
        r += (color.r * numerator) >> 4;
        g += (color.g * numerator) >> 4;
        b += (color.b * numerator) >> 4;
        return this;
    }

    public SplitColor sub(SplitColor color) {
        return new SplitColor(
                r - color.r,
                g - color.g,
                b - color.b
        );
    }

    public double diff(SplitColor color) {
        return Math.pow(r - color.r, 2)
                + Math.pow(g - color.g, 2)
                + Math.pow(b - color.b, 2);
    }

    @ColorInt
    public int toColor() {
        return Color.argb(
                0xff,
                ColorUtil.clampChannel(r),
                ColorUtil.clampChannel(g),
                ColorUtil.clampChannel(b)
        );
    }
}

File: app/src/main/java/de/troido/bledemo/epd/conversion/SplitColor.kt
package de.troido.bledemo.epd.conversion

import android.graphics.Color
import androidx.annotation.ColorInt
import kotlin.math.pow

/**
 * Color split into channels which can take up negative values for dithering
 * error purposes.
 */
class SplitColor(private var r: Int, private var g: Int, private var b: Int) {

    constructor(@ColorInt color: Int) : this(Color.red(color), Color.green(color), Color.blue(color)) {}

    /**
     * Adds the given error color multiplied by the given numerator
     * (with 16 as denominator), mutating this object.
     */
    fun addErrMut(color: SplitColor, numerator: Int): SplitColor {
        r += color.r * numerator shr 4
        g += color.g * numerator shr 4
        b += color.b * numerator shr 4
        return this
    }

    fun sub(color: SplitColor): SplitColor {
        return SplitColor(
            r - color.r,
            g - color.g,
            b - color.b
        )
    }

    fun diff(color: SplitColor): Double {
        return ((r - color.r.toDouble()).pow(2.0)
            + (g - color.g.toDouble()).pow(2.0)
            + (b - color.b.toDouble()).pow(2.0))
    }

    @ColorInt
    fun toColor(): Int {
        return Color.argb(
            0xFF,
            ColorUtil.clampChannel(r),
            ColorUtil.clampChannel(g),
            ColorUtil.clampChannel(b)
        )
    }

}